staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -251,16 +254,89 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
 			continue
 		}
 		if evalResult != types.True {
-			if len(rule.Message) != 0 {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, rule.Message))
+			if compiled.MessageExpression != nil {
+				messageExpression, newRemainingBudget, msgErr := evalMessageExpression(ctx, compiled.MessageExpression, rule.MessageExpression, activation, remainingBudget)
+				if msgErr != nil {
+					if msgErr.errType == runtimeCostErr {
+						errs = append(errs, field.InternalError(fldPath, msgErr.err))
+						continue
+					} else {
+						klog.Warning(msgErr.err)
+						errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
+					}
+				} else {
+					errs = append(errs, field.Invalid(fldPath, sts.Type, messageExpression))
+					remainingBudget = newRemainingBudget
+				}
 			} else {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, fmt.Sprintf("failed rule: %s", ruleErrorString(rule))))
+				errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
```

Comment:
RuleErrorString here already handles using message or rule as returned message back to user right?

Reply:
I think this is the same case as [here](https://github.com/kubernetes/kubernetes/pull/115969#discussion_r1126804627). We could remove `ruleMessageOrDefault`, but I think we'd have to duplicate the logic here twice: https://github.com/kubernetes/kubernetes/blob/0270fc75d0f59e1f5a31a810737ac35ac3eeb8a3/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go#L254-L258

Reply:
I see. This makes sense for having `failed rule:` only when returning rule as message. :)